File: plugins/haptics/android/src/main/java/HapticsPlugin.kt
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

package app.tauri.haptics

import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import app.tauri.Logger
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.haptics.patterns.ImpactPatternHeavy
import app.tauri.haptics.patterns.ImpactPatternLight
import app.tauri.haptics.patterns.ImpactPatternMedium
import app.tauri.haptics.patterns.ImpactPatternRigid
import app.tauri.haptics.patterns.ImpactPatternSoft
import app.tauri.haptics.patterns.NotificationPatternError
import app.tauri.haptics.patterns.NotificationPatternSuccess
import app.tauri.haptics.patterns.NotificationPatternWarning
import app.tauri.haptics.patterns.Pattern
import app.tauri.haptics.patterns.SelectionPattern
import app.tauri.plugin.Invoke
import app.tauri.plugin.Plugin
import com.fasterxml.jackson.annotation.JsonProperty

@InvokeArg
class HapticsOptions {
  var duration: Long = 300
}

@InvokeArg
class NotificationFeedbackArgs {
    val type: NotificationFeedbackType = NotificationFeedbackType.Success
}

@InvokeArg
enum class NotificationFeedbackType {
    @JsonProperty("success")
    Success,
    @JsonProperty("warning")
    Warning,
    @JsonProperty("error")
    Error;

    fun into(): Pattern {
        return when(this) {
            Success -> NotificationPatternSuccess
            Warning -> NotificationPatternWarning
            Error -> NotificationPatternError
        }
    }
}

@InvokeArg
class ImpactFeedbackArgs {
    val style: ImpactFeedbackStyle = ImpactFeedbackStyle.Medium
}

@InvokeArg
enum class ImpactFeedbackStyle {
    @JsonProperty("light")
    Light,
    @JsonProperty("medium")
    Medium,
    @JsonProperty("heavy")
    Heavy,
    @JsonProperty("soft")
    Soft,
    @JsonProperty("rigid")
    Rigid;

    fun into(): Pattern {
        return when(this) {
            Light -> ImpactPatternLight
            Medium -> ImpactPatternMedium
            Heavy -> ImpactPatternHeavy
            Soft -> ImpactPatternSoft
            Rigid -> ImpactPatternRigid
        }
    }
}

@TauriPlugin
class HapticsPlugin(private val activity: Activity): Plugin(activity) {
    private val vibrator: Vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        val vibManager = activity.applicationContext.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
        vibManager.defaultVibrator
    } else {
        @Suppress("DEPRECATION")
        activity.applicationContext.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
    }

    //
    // TAURI COMMANDS
    //

    @Command
    fun vibrate(invoke: Invoke) {
        val args = invoke.parseArgs(HapticsOptions::class.java)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(args.duration, VibrationEffect.DEFAULT_AMPLITUDE))
        } else {
            vibrator.vibrate(args.duration)
        }
        invoke.resolve()
    }

    @Command
    fun impactFeedback(invoke: Invoke) {
        val args = invoke.parseArgs(ImpactFeedbackArgs::class.java)
        vibratePattern(args.style.into())
        invoke.resolve()
    }

    @Command
    fun notificationFeedback(invoke: Invoke) {
        val args = invoke.parseArgs(NotificationFeedbackArgs::class.java)
        vibratePattern(args.type.into())
        invoke.resolve()
    }

    // TODO: Consider breaking this up into Start,Change,End like capacitor
    @Command
    fun selectionFeedback(invoke: Invoke) {
        vibratePattern(SelectionPattern)
        invoke.resolve()
    }

    // INTERNAL FUNCTIONS

    private fun vibratePattern(pattern: Pattern) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createWaveform(pattern.timings, pattern.amplitudes, -1))
        } else {
            vibrator.vibrate(pattern.oldSDKPattern, -1)
        }
    }
}

File: plugins/haptics/ios/Sources/HapticsPlugin.swift
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

import SwiftRs
import Tauri
import UIKit
import WebKit
import CoreHaptics
import AudioToolbox

class ImpactFeedbackOptions: Decodable {
  let style: ImpactFeedbackStyle
}

enum ImpactFeedbackStyle: String, Decodable {
  case light, medium, heavy, soft, rigid

  func into() -> UIImpactFeedbackGenerator.FeedbackStyle {
    switch self {
    case .light:
      return .light
    case .medium:
      return .medium
    case .heavy:
      return .heavy
    case .soft:
      return .soft
    case .rigid:
      return .rigid
    }
  }
}

class NotificationFeedbackOptions: Decodable {
  let type: NotificationFeedbackType
}

enum NotificationFeedbackType: String, Decodable {
  case success, warning, error

  func into() -> UINotificationFeedbackGenerator.FeedbackType {
    switch self {
    case .success:
      return .success
    case .warning:
      return .warning
    case .error:
      return .error
    }
  }
}

class VibrateOptions: Decodable {
  // TODO: Array
  let duration: Double
}

class HapticsPlugin: Plugin {
  //
  // Tauri commands
  //

  @objc public func vibrate(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(VibrateOptions.self)
    if CHHapticEngine.capabilitiesForHardware().supportsHaptics {
      do {
        let engine = try CHHapticEngine()
        try engine.start()
        engine.resetHandler = { [] in
          do {
              try engine.start()
          } catch {
            AudioServicesPlayAlertSound(kSystemSoundID_Vibrate)
          }
        }
        // TODO: Make some of this (or all) configurable?
        let intensity: CHHapticEventParameter = CHHapticEventParameter(parameterID: .hapticIntensity, value: 1.0)
        let sharpness: CHHapticEventParameter = CHHapticEventParameter(parameterID: .hapticSharpness, value: 1.0)
        let continuousEvent = CHHapticEvent(
          eventType: .hapticContinuous,
          parameters: [intensity, sharpness],
          relativeTime: 0.0,
          duration: args.duration/1000
        )
        let pattern = try CHHapticPattern(events: [continuousEvent], parameters: [])
        let player = try engine.makePlayer(with: pattern)

        try player.start(atTime: 0)
      } catch {
        AudioServicesPlayAlertSound(kSystemSoundID_Vibrate)
      }
    } else {
      AudioServicesPlayAlertSound(kSystemSoundID_Vibrate)
    }

    Logger.error("VIBRATE END")

    invoke.resolve()
  }

  @objc public func impactFeedback(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(ImpactFeedbackOptions.self)
    let generator = UIImpactFeedbackGenerator(style: args.style.into())
    generator.prepare()
    generator.impactOccurred()

    invoke.resolve()
  }

  @objc public func notificationFeedback(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(NotificationFeedbackOptions.self)
    let generator = UINotificationFeedbackGenerator()
    generator.prepare()
    generator.notificationOccurred(args.type.into())

    invoke.resolve()
  }

  // TODO: Consider breaking this up into Start,Change,End like capacitor
  @objc public func selectionFeedback(_ invoke: Invoke) throws {
    let generator = UISelectionFeedbackGenerator()
    generator.prepare()
    generator.selectionChanged()

    invoke.resolve()
  }
}

@_cdecl("init_plugin_haptics")
func initPlugin() -> Plugin {
  return HapticsPlugin()
}
